semgrep-core/src/parsing/tree_sitter/Parse_hcl_tree_sitter.ml

```diff
@@ -85,16 +85,38 @@ let map_numeric_lit (env : env) (x : CST.numeric_lit) : literal =
 
 let map_template_literal (env : env) (xs : CST.template_literal) :
     string wrap option =
-  (* each character is a separate template_literal_chunk, so we need
-   * to merge them in a single string.
-   *)
-  let xs = List.map (str env (* template_literal_chunk *)) xs in
   match xs with
   | [] -> None
-  | (s1, t1) :: ys ->
-      let str = List.map fst ys |> String.concat "" in
-      let toks = List.map snd ys in
-      Some (s1 ^ str, PI.combine_infos t1 toks)
+  | ((x_pos, x_str) as x) :: xs ->
+      (* each character is a separate template_literal_chunk, so we need
+       * to merge them in a single string. *)
+      let module Loc = Tree_sitter_run.Loc in
+      let base_col = x_pos.Loc.start.column in
+      let rec concat_chunks acc_str prev_row prev_col xs =
+        match xs with
+        | [] -> acc_str
+        | x :: xs ->
+            let x_loc, x_str = x in
+            let last_col =
+              if x_loc.Loc.start.row > prev_row then base_col else prev_col
+            in
+            let acc_str =
+              (* we need to respect newlines and spaces so that HEREDOCs are
```

Comment:
This is great, might be better as a function comment so a reader can understand what map_template_string needs to do without looking into the code